### ISSUE TITLE
Transaction cannot detect commit failure

### ISSUE THREAD

**Issue by artsince** [neo4jrb/neo4j#1001]
I am running some cypher queries inside the Transaction block. Even though neo4j-core gem thinks the transaction ends successfully, I check the http logs of neo4j and I see that the commit has failed.

What I am specifically trying to do is to trace a bug where i batch insert some data inside a transaction but somehow the transaction is rolled back without any error notification. After digging around the transaction logs I find the error I pasted below at the end of the transaction period.

I am basically sending about 800 cypher queries in a transaction like this and no exception is thrown:
```rb
    begin
      Neo4j::Transaction.run do
       # a lot of Neo4j::Session.query("")
      end
    rescue Exception => exp
      puts exp.message
      puts exp.backtrace.join("\n")
      return false
    end
```

I also try the other approach, and check the failed? flag, no exception is thrown
```rb
    begin
        tx = Neo4j::Transaction.new
        # a lot of Neo4j::Session.query("")
    rescue Exception => exp
      puts exp.message
      puts exp.backtrace.join("\n")
      tx.mark_failed
      return false
    ensure
      puts "TRANSACTION FAILED = " + tx.failed?.to_s
      tx.close
    end
```

Each query creates a request like this:
```
127.0.0.1 - - [22/Oct/2015:13:43:48 +0300] "POST /db/data/transaction/5 HTTP/1.1" 200 172 "-" "neo4j-gem/5.2.3 (https://github.com/neo4jrb/neo4j)" 
Request:
POST /db/data/transaction/5 HTTP/1.1
Host: localhost:7474
Content-Length: 981
Accept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3
User-Agent: neo4j-gem/5.2.3 (https://github.com/neo4jrb/neo4j)
Keep-Alive: 30
Content-Type: application/json;charset=UTF-8
Accept: */*
Connection: keep-alive

{"statements":[{"statement":"MATCH (g:Graph {_id: 'a84f7921-13bd-4b9d-85b5-4cead6980738'})\n\nMATCH (o:User {uuid: 'd012995d-2e9b-43b9-a1a6-677fb4ef8f40'})\n\nMATCH (et:EdgeType {_id: 'b8d886a2-5876-6bbe-57fa-471001140225'})\nMATCH (fromNode:Node {_id: 'cf57cdd8-14d7-4552-9bc8-e58434677410'})\nMATCH (toNode:Node {_id: 'acc79123-3161-4b38-9379-50787676fd80'})\nMERGE (fromNode)-[r:`CONTRACT`]-\u003e(toNode)\nON CREATE \nSET r.client_id  = 'cd007087-c26e-4d38-abb9-6dcec4acff18',\n    r.type_id    = 'b8d886a2-5876-6bbe-57fa-471001140225',\n    r.graph_ids  = ['a84f7921-13bd-4b9d-85b5-4cead6980738'],\n    r.hub_id     = NULL,\n    r.user_id    = 'd012995d-2e9b-43b9-a1a6-677fb4ef8f40',\n    r.weight     = 1,\n    r.directed   = 1,\n    r.properties = '{}',\n    et.edges_count = COALESCE(et.edges_count, 0) + 1\nCREATE UNIQUE (toNode)-[:EDGE_TYPES_IN]-\u003e(et)\nCREATE UNIQUE (fromNode)-[:EDGE_TYPES_OUT]-\u003e(et)\n","parameters":null,"resultDataContents":["row","REST"]}]}
Response:
HTTP/1.1 200 OK
Date: Thu, 22 Oct 2015 10:43:48 GMT
Access-Control-Allow-Origin: *
Content-Type: application/json

{"commit":"http://localhost:7474/db/data/transaction/5/commit","results":[{"columns":[],"data":[]}],"transaction":{"expires":"Thu, 22 Oct 2015 10:44:48 +0000"},"errors":[]}
```

and the final commit request is:
```
127.0.0.1 - - [22/Oct/2015:13:43:49 +0300] "POST /db/data/transaction/5/commit HTTP/1.1" 200 5045 "-" "neo4j-gem/5.2.3 (https://github.com/neo4jrb/neo4j)" 
Request:
POST /db/data/transaction/5/commit HTTP/1.1
Host: localhost:7474
Content-Length: 0
Accept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3
Keep-Alive: 30
User-Agent: neo4j-gem/5.2.3 (https://github.com/neo4jrb/neo4j)
Content-Type: application/json;charset=UTF-8
Accept: application/json; charset=UTF-8
Connection: keep-alive


Response:
HTTP/1.1 200 OK
Date: Thu, 22 Oct 2015 10:43:48 GMT
Access-Control-Allow-Origin: *
Content-Type: application/json

{"results":[],"errors":[{"code":"Neo.DatabaseError.Transaction.CouldNotCommit","message":"3486144 already exists","stackTrace":"java.lang.IllegalStateException: 3486144 already exists\n\tat org.neo4j.kernel.impl.transaction.state.RecordChanges.create(RecordChanges.java:116)\n\tat org.neo4j.kernel.impl.transaction.state.PropertyCreator.addPropertyBlockToPrimitive(PropertyCreator.java:162)\n\tat org.neo4j.kernel.impl.transaction.state.PropertyCreator.primitiveAddProperty(PropertyCreator.java:124)\n\tat org.neo4j.kernel.impl.transaction.state.NeoStoreTransactionContext.primitiveAddProperty(NeoStoreTransactionContext.java:95)\n\tat org.neo4j.kernel.impl.transaction.state.TransactionRecordState.nodeAddProperty(TransactionRecordState.java:377)\n\tat org.neo4j.kernel.impl.api.KernelTransactionImplementation$TransactionToRecordStateVisitor.visitNodePropertyChanges(KernelTransactionImplementation.java:796)\n\tat org.neo4j.kernel.impl.api.state.TxState$18.visitPropertyChanges(TxState.java:529)\n\tat org.neo4j.kernel.impl.api.state.PropertyContainerState$Mutable.accept(PropertyContainerState.java:284)\n\tat org.neo4j.kernel.impl.api.state.NodeState$Mutable.accept(NodeState.java:305)\n\tat org.neo4j.kernel.impl.api.state.TxState.accept(TxState.java:285)\n\tat org.neo4j.kernel.impl.api.KernelTransactionImplementation.prepareRecordChangesFromTransactionState(KernelTransactionImplementation.java:395)\n\tat org.neo4j.kernel.impl.api.KernelTransactionImplementation.commit(KernelTransactionImplementation.java:528)\n\tat org.neo4j.kernel.impl.api.KernelTransactionImplementation.close(KernelTransactionImplementation.java:456)\n\tat org.neo4j.server.rest.transactional.TransitionalTxManagementKernelTransaction.commit(TransitionalTxManagementKernelTransaction.java:83)\n\tat org.neo4j.server.rest.transactional.TransactionHandle.closeContextAndCollectErrors(TransactionHandle.java:279)\n\tat org.neo4j.server.rest.transactional.TransactionHandle.commit(TransactionHandle.java:148)\n\tat org.neo4j.server.rest.web.TransactionalService$2.write(TransactionalService.java:211)\n\tat com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:71)\n\tat com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:57)\n\tat com.sun.jersey.spi.container.ContainerResponse.write(ContainerResponse.java:302)\n\tat com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1510)\n\tat com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419)\n\tat com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409)\n\tat com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409)\n\tat com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:558)\n\tat com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:733)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:790)\n\tat org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:800)\n\tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)\n\tat ch.qos.logback.access.servlet.TeeFilter.doFilter(TeeFilter.java:55)\n\tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)\n\tat org.neo4j.server.rest.web.CollectUserAgentFilter.doFilter(CollectUserAgentFilter.java:69)\n\tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)\n\tat org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)\n\tat org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)\n\tat org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1125)\n\tat org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)\n\tat org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)\n\tat org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1059)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)\n\tat org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)\n\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)\n\tat org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:95)\n\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)\n\tat org.eclipse.jetty.server.Server.handle(Server.java:497)\n\tat org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)\n\tat org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:248)\n\tat org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:620)\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:540)\n\tat java.lang.Thread.run(Thread.java:745)\n"}]}
```

I understand that the underlying problem I am trying to solve is somewhere else, but my inability to detect this rollback prevents me from handling this error case gracefully.

I am using neo4j-gem/5.2.3 and neo4j-2.3.0. I got the same error when I upgraded to 5.2.11.

What do you think? Is there another approach I should take?

**Comment by subvertallchris** [comment]
Can you try with Neo4j 2.2.6? We haven't tested the gems with 2.3.0
yet and it's
possible that a big or change there is responsible.

What happens if you commit more often? 800 queries might be on the heavy
side. If you can, drop it down low, maybe 50 queries, and gradually
increase the number if it works. Shortly after we released the 4.0 version
of the gem, I was experimenting with a migration task to add UUIDs and
found that if I committed too infrequently, I'd run into memory issues that
resulted in commit failures, so maybe that's what is happening here.

**Comment by artsince** [comment]
hey, I tried with 2.2.5 before switching to 2.3, I will try with 2.2.6 soon. I actually try with smaller batches like you suggested, I get inconsistent behavior. (Rolling back the first time, but committing the second time.)
In any case, is there a way that neo4rb can catch these CouldNotCommit errors and notify me of the rollback?

**Comment by subvertallchris** [comment]
I see that it's returning 200s with those errors. In Neo4j-core, we only raise errors when the response code is something other than 200, so I think that might be the problem. It's odd, though, because we've had it setup like that for a long time without issue and I know that I see errors all the time, but Neo4j's documentation says that a 200 response with `error` key is normal.

I'm working on something right now but as soon as I finish up, I'll take a look. Is there any chance you can link me to the code that lets you reproduce this?

**Comment by artsince** [comment]
thanks. I will see if I can reproduce this error on a smaller, isolated project.

I will also post an issue on the neo4j repository to see if they can help me with the actual error I am trying to solve: `"message":"3486144 already exists"`

**Comment by cheerfulstoic** [comment]
Just added 2.3.0 to the Travis builds for neo4j and neo4j-core

**Comment by artsince** [comment]
Hi @subvertallchris, I was unable to reproduce this problem in a smaller project with a clean database. It seems like my current database is in a faulty state after a restored back up.

**Comment by subvertallchris** [comment]
Got it, thanks for following up. Can we close this issue?

**Comment by artsince** [comment]
While my problem is resolved, I still think this is a case that needs to be handled

According to the documentation [here](http://neo4j.com/docs/stable/rest-api-transactional.html#rest-api-handling-errors)
> all requests against the transactional endpoint will return 200 or 201 status code, regardless of whether statements were successfully executed.

So, I think the transaction should be set as failed when there is an error like this.

Thanks,

**Comment by subvertallchris** [comment]
Yup, I'm working on that right now. Forgot I had commented earlier. I'll update here as soon as I have some news.

**Comment by subvertallchris** [comment]
Ok, I see how this is supposed to work.

TL;DR The `CypherResponse` class is supposed to look for the error and raises later on.

`Neo4j::Server::CypherTransaction#_query` ends with a call to `create_cypher_response` (https://github.com/neo4jrb/neo4j-core/blob/master/lib/neo4j-server/cypher_transaction.rb#L30), which does a check for `response.body[:errors]`. If errors are found, it adds details about the error to the response. At different points, `response.raise_error if response.error?` appears and if one of them catches the error, it explodes.

We're in the process of rebuilding this area of Neo4j-core, so that process and its code will be cleaned up soon. In the meantime, is it possible that something in your code is preventing it from showing up as expected?

**Comment by subvertallchris** [comment]
This was fixed in recent versions of both 6.0.0.alpha and 5.2, so I'm closing this. Let us know if you have any trouble.
